streamer/src/nonblocking/quic.rs

```diff
@@ -186,35 +187,55 @@ async fn handle_connection(
         stats.total_connections.load(Ordering::Relaxed),
     );
     while !stream_exit.load(Ordering::Relaxed) {
-        match uni_streams.next().await {
-            Some(stream_result) => match stream_result {
-                Ok(mut stream) => {
-                    stats.total_streams.fetch_add(1, Ordering::Relaxed);
-                    stats.total_new_streams.fetch_add(1, Ordering::Relaxed);
-                    let mut maybe_batch = None;
-                    while !stream_exit.load(Ordering::Relaxed) {
-                        if handle_chunk(
-                            &stream.read_chunk(PACKET_DATA_SIZE, false).await,
-                            &mut maybe_batch,
-                            &remote_addr,
-                            &packet_sender,
-                            stats.clone(),
-                            stake,
-                        ) {
-                            last_update.store(timing::timestamp(), Ordering::Relaxed);
-                            break;
+        if let Ok(stream) = tokio::time::timeout(
+            Duration::from_millis(WAIT_FOR_STREAM_TIMEOUT_MS),
+            uni_streams.next(),
+        )
+        .await
+        {
+            match stream {
+                Some(stream_result) => match stream_result {
+                    Ok(mut stream) => {
+                        stats.total_streams.fetch_add(1, Ordering::Relaxed);
+                        stats.total_new_streams.fetch_add(1, Ordering::Relaxed);
+                        let mut maybe_batch = None;
+                        while !stream_exit.load(Ordering::Relaxed) {
+                            if let Ok(chunk) = tokio::time::timeout(
+                                Duration::from_millis(WAIT_FOR_STREAM_TIMEOUT_MS),
+                                stream.read_chunk(PACKET_DATA_SIZE, false),
+                            )
+                            .await
+                            {
+                                if handle_chunk(
+                                    &chunk,
+                                    &mut maybe_batch,
+                                    &remote_addr,
+                                    &packet_sender,
+                                    stats.clone(),
+                                    stake,
+                                ) {
+                                    last_update.store(timing::timestamp(), Ordering::Relaxed);
+                                    break;
+                                }
+                            } else {
+                                debug!("Timeout in receiving on stream");
+                                stats
+                                    .total_stream_read_timeouts
+                                    .fetch_add(1, Ordering::Relaxed);
+                                break;
```

Comment:
This break means that we abandon a stream if it takes longer than one second to receive the chunk. That feels too short to me, do we need to abandon the stream?

The timeout you added is good for ensuring that we check `stream_exit` often enough but that doesn't require abandoning the stream.

Reply:
In our use case of streams, we are creating stream, sending transaction and closing the stream in one shot. So if the stream got created, but we didn't receive enough chunks, it's probably better to abandon it.

Reply:
That's not exactly what I was referring to. I believe the tx can be sent in a separate UDP datagram which could be dropped and need to be retried by the client. I think that adding a 1s timeout makes it less likely that transactions will be delivered when there's lots of network traffic